## 1.3.1 Procedures as Arguments

### Example 1: summation

Consider the following three procedures.

The first computes the sum of the integers from a through b:

The second computes the sum of the cubes of the integers in the given range:

The third computes the sum of a sequence of terms in the series  

$$\frac{1}{1 \centerdot 3} + \frac{1}{5 \centerdot 7} + \frac{1}{9 \centerdot 11} + \ldots$$

which converges to $\pi$/8 (very slowly):

Indeed, mathematicians long ago identified the abstraction of summation of a series and invented "sigma notation", for example,to express this concept.

$$\sum_{n=a}^bf(n) = f(a) + \ldots+ f(b)$$

With the aid of the procedures sum as follows,we can express the mathematical concepts of summation:

So,we can redefine the process sum-integers to compute the sum of the integers from a through b:

In [1]:
cat 1.3/sum_integers_for_example.scm

(load "1.3/sum_by_recursion.scm")

(define (identity x) x)
(define (inc n) (+ n 1))

(define (sum-integers a b)
  (sum identity a inc b))


 Redefine the process sum-cubes to compute the sum of the cubes of the integers from a through b:

In [2]:
cat 1.3/sum_cubes_for_example.scm

(load "1.3/sum_by_recursion.scm")

(define (cube x) (* x x x))
(define (inc n) (+ n 1))

(define (sum-cubes a b)
  (sum cube a inc b))


We can also redefine pi-sum in the same way:

In [5]:
cat 1.3/sum_pi_by_recursion.scm

(load "1.3/sum_by_recursion.scm")

(define (pi-sum a b)
  (define (pi-term x)
    (/ 1.0 (* x (+ x 2))))
  (define (pi-next x) (+ x 4))
  (sum pi-term a pi-next b))


### Example 2: numerical integration

Once we have sum, we can use it as a building block in formulating further concepts. For instance, the definite integral of a function f between the limits a and b can be approximated numerically using the formula

$$\int_a^bf=\left[f\left(a + \frac{dx}{2}\right) + f\left(a + dx + \frac{dx}{2}\right) + f\left(a + 2dx + \frac{dx}{2}\right) + \centerdot\centerdot\centerdot\right]dx$$

for small values of dx. We can express this directly as a procedure:

In [5]:
cat 1.3/compute_integral_for_function.scm

(load "1.3/sum_by_recursion.scm")

(define (integral f a b dx)
  (define (add-dx x) (+ x dx))
  (* (sum f (+ a (/ dx 2.0)) add-dx b) dx))


For example,we can let f is cube function:

### Exercise 1.29:
Simpson’s Rule is a more accurate method of numerical integration than the method illustrated above. Using Simpson’s Rule, the integral of a function f between a and b is approximated as

$$\frac{h}{3}(y_0 + 4y_1 + 2y_2 + 4y_3 + 2y_4 + \cdots + 2y_{n-2} + 4y_{n-1} + y_n)$$

where $h = (b -a)/n$,for some even integer n, and $y_k = f(a + kh)$.(Increasing n increases the accuracy of the approximation.)Define a procedure that takes as arguments $f$ , $a$, $b$, and $n$ and returns the value of the integral, computed using Simpson’s Rule.Use your procedure to integrate cube between 0 and 1 (with $n = 100$ and $n = 1000$), and compare the results to those of the integral procedure shown above.

### Exercise 1.30: 
The sum procedure above generates a linear recursion. The procedure can be rewrien so that the sum is performed iteratively. Show how to do this by filling in the missing expressions in the following definition:

### Answer:

In [1]:
cat 1.3/sum_by_iteration.scm

(define (sum term a next b)
  (define (iter a result)
    (if (> a b)
        result
        (iter (next a)
              (+ (term a) result))))
  (iter a 0))
		


1.Compute the sum of the integers from a to b:

In [2]:
cat 1.3/sum_integers_for_exercise130.scm

(load "1.3/sum_by_iteration.scm")

(define (identity x) x)
(define (inc n) (+ n 1))

(define (sum-integers a b)
  (sum identity a inc b))


2.Compute the sum of the cubes of the integers from a to b:

In [3]:
cat 1.3/sum_cubes_for_exercise130.scm

(load "1.3/sum_by_iteration.scm")

(define (cube x) (* x x x))
(define (inc n) (+ n 1))

(define (sum-cubes a b)
  (sum cube a inc b))


3.Compute the sum-pi procedure by way of iterative process: 

In [4]:
cat 1.3/sum_pi_for_exercise130.scm

(load "1.3/sum_by_iteration.scm")

(define (pi-sum a b)
  (define (pi-term x)
    (/ 1.0 (* x (+ x 2))))
  (define (pi-next x) (+ x 4))
  (sum pi-term a pi-next b))


### Exercise 1.31:
a. The $sum$ procedure is only the simplest of a vast number of similar abstractions that can be captured as higher-order procedures.Write an analogous procedure called $product$ that returns the product of the values of a
function at points over a given range. Show how to define $factorial$ in terms of $product$. Also use product to compute approximations to $\pi$ using the formula

$$\frac{\pi}{4} = \frac{2\centerdot4\centerdot4\centerdot6\centerdot6\centerdot8\cdots}{3\centerdot3\centerdot5\centerdot5\centerdot7\centerdot7\cdots}$$

### Answer:

1.The product procedure as follows.

In [1]:
cat 1.3/product_by_recursion.scm

(define (product term a next b)
  (if (> a b) 1
      (* (term a)
         (product term (next a) next b))))


2.Redefine factorial procedure in terms of product procedure.

In [2]:
cat 1.3/factorial_by_product_by_recursion.scm

(load "1.3/product_by_recursion.scm")

(define (identity x) x)
(define (inc n) (+ n 1))

(define (factorial b)
  (product identity 1 inc b))


3.Use product to compute approximations to $\pi$.

In [3]:
cat 1.3/product_pi_by_recursion.scm

(load "1.3/product_by_recursion.scm")

(define (pi-next n) (+ n 2))
(define (pi-term a)
  ( * 1.0 (/ (* a (pi-next a)) (square (+ a 1)))))

(define (product-pi a b)
  (product pi-term a pi-next b))


b. If your product procedure generates a recursive process,write one that generates an iterative process. If it generates an iterative process, write one that generates a recursive process.

### Answer:

Since the product procedure above generates a recursive process;we should write one that generates an iterative process

1.The product procedure as follows.

In [5]:
cat 1.3/product_by_iteration.scm

(define (product term a next b)
  (define (iter a result)
    (if (> a b) result
        (iter (next a)
              (* (term a) result))))
  (iter a 1))


Test:

2.Redefine factorial procedure

In [6]:
cat 1.3/factorial_by_product_by_iteration.scm

(load "1.3/product_by_iteration.scm")

(define (identity x) x)
(define (inc n) (+ n 1))

(define (factorial b)
  (product identity 1 inc b))


3.Use product to compute approximations to $\pi$

In [7]:
cat 1.3/product_pi_by_iteration.scm

(load "1.3/product_by_iteration.scm")

(define (pi-next n) (+ n 2))
(define (pi-term a)
  ( * 1.0 (/ (* a (pi-next a)) (square (+ a 1)))))

(define (product-pi a b)
  (product pi-term a pi-next b))


### Exercise 1.32:
#### a. 
Show that $sum$ and $product$ (Exercise 1.31) are both special cases of a still more general notion called $accumulate$ that combines a collection of terms, using some general accumulation function:

$accumulate$ takes as arguments the same term and range specifications as $sum$ and $product$, together with a $combiner$ procedure (of two arguments) that specifies how the current term is to be combined with the accumulation of the preceding terms and a null-value that specifies what base value to use when the terms run out. Write $accumulate$ and show how $sum$ and $product$ can both be defined as simple calls to $accumulate$.

1.The $accumulate$ recursive procedure:

In [8]:
cat 1.3/accumulate_by_recursion.scm

(define (accumulate combiner null-value term a next b)
  (if (> a b) null-value
      (combiner (term a)
                (accumulate combiner
                            null-value
                            term
                            (next a)
                            next
                            b))))


2.Redefine the $sum$ procedure by way of the $accumulate$ recursive procedure:

In [11]:
cat 1.3/sum_by_accumulate_by_recursion.scm

(load "1.3/accumulate_by_recursion.scm")

(define (sum term a next b)
  (accumulate + 0 term a next b))


3.Redefine the $product$ procedure by way of the $accumulate$ recursive procedure:

In [14]:
cat 1.3/product_by_accumulate_by_recursion.scm

(load "1.3/accumulate_by_recursion.scm")

(define (product term a next b)
  (accumulate * 1 term a next b))


#### Test:

1.summation

In [20]:
cat 1.3/sum_integers_by_accumulate_by_recursion.scm

(load "1.3/sum_by_accumulate_by_recursion.scm")

(define (identity x) x)
(define (inc n) (+ n 1))

(define (sum-integers a b)
  (sum identity a inc b))


2.product

In [22]:
cat 1.3/product_pi_by_accumulate_by_recursion.scm

(load "1.3/product_by_accumulate_by_recursion.scm")

(define (pi-next n) (+ n 2))
(define (pi-term a)
    (* 1.0 (/ (* a (pi-next a)) (square (+ a 1)))))

(define (product-pi a b)
    (product pi-term a pi-next b))


#### b. 
If your $accumulate$ procedure generates a recursive process, write one that generates an iterative process.If it generates an iterative process, write one that generates a recursive process.

1.The accumulate iterative procedure:

In [10]:
cat 1.3/accumulate_by_iteration.scm

(define (accumulate combiner null-value term a next b)
  (define (iter a result)
    (if (> a b) result
        (iter (next a)
              (combiner (term a) result))))
  (iter a null-value))


2.Redefine the sum procedure by way of the accumulate iterative procedure:

In [13]:
cat 1.3/sum_by_accumulate_by_iteration.scm

(load "1.3/accumulate_by_iteration.scm")

(define (sum term a next b)
  (accumulate + 0 term a next b))


3.Redefine the $product$ procedure by way of the $accumulate$ iterative procedure:

In [15]:
cat 1.3/product_by_accumulate_by_iteration.scm

(load "1.3/accumulate_by_iteration.scm")

(define (product term a next b)
  (accumulate * 1 term a next b))


#### Test:

1.summation

In [21]:
cat 1.3/sum_integers_by_accumulate_by_iteration.scm

(load "1.3/sum_by_accumulate_by_iteration.scm")

(define (identity x) x)
(define (inc n) (+ n 1))

(define (sum-integers a b)
  (sum identity a inc b))


2.product

In [23]:
cat 1.3/product_pi_by_accumulate_by_iteration.scm

(load "1.3/product_by_accumulate_by_iteration.scm")

(define (pi-next n) (+ n 2))
(define (pi-term a)
    ( * 1.0 (/ (* a (pi-next a)) (square (+ a 1)))))

(define (product-pi a b)
    (product pi-term a pi-next b))


### Exercise 1.33:
You can obtain an even more general version of $accumulate$ (Exercise 1.32) by introducing the notion of a $filter$ on the terms to be combined. That is, combine only those terms derived from values in the range that satisfy a specified condition. The resulting $filtered$-$accumulate$ abstraction takes the same arguments as $accumulate$, together with an additional predicate of one argument that specifies the $filter$. Write filtered accumulate as a procedure.Show how to express the following using $filtered$-$accumulate$:

### Answer:

#### A.The filtered-accumulate recursive procedure:

In [24]:
cat 1.3/filtered_accumulate_by_recursion.scm

(define (filtered-accumulate combine null-value term a next b valid)
  (if (> a b) null-value
      (if (valid a)
         (combine (term a)
                  (filtered-accumulate combine
                                       null-value
                                       term
                                       (next a)
                                       next
                                       b
                                       valid))
        (filtered-accumulate combine
       	                     null-value
       	                     term
       	                     (next a)
       	                     next
       	                     b
       	                     valid))))


#### a. 
the $sum$ of the squares of the prime numbers in the interval a to b (assuming that you have a $prime$ predicate already written)

In [25]:
cat 1.3/sum_square_of_prime_by_filtered_accumulate_by_recursion.scm

(load "1.3/filtered_accumulate_by_recursion.scm")
(load "1.2/prime_by_search_divisor1.scm")

(define (inc n) (+ n 1))

(define (sum-prime a b)
  (filtered-accumulate + 0 square a inc b prime))


#### b. 
the $product$ of all the positive integers less than n that are relatively prime to n  
(i.e., all positive integers $i < n$ such that $GCD(i,n) = 1$).

In [26]:
cat 1.3/product_of_coprimes.scm

(load "1.3/filtered_accumulate_by_recursion.scm")
(load "1.2/gcd_by_Euclid_Algorithm.scm")

(define (identity x) x)
(define (coprime a b) (and (< a b) (= 1 (gcd a b))))
(define (inc n) (+ n 1))

(define (product-of-coprimes n)
  (filtered-accumulate * 1 identity 1 inc n (lambda (x) (coprime x n))))


#### B.The filtered-accumulate iterative procedure:

In [1]:
cat 1.3/filtered_accumulate_by_iteration.scm

(define (filtered-accumulate combine null-value term a next b valid)
  (define (iter i result)
    (cond ((> i b) result)
          ((valid i)
           (iter (next i) (combine (term i) result)))
          (else (iter (next i) result))))
  (iter a null-value))


#### a.
the sum of the cubes of the prime numbers in the interval a to b:

In [2]:
cat 1.3/sum_cube_of_prime_by_filtered_accumulate_by_iteration.scm

(load "1.3/filtered_accumulate_by_iteration.scm")
(load "1.2/prime_by_search_divisor1.scm")

(define (inc n) (+ n 1))
(define (cube x) (* x x x))

(define (sum-prime a b)
  (filtered-accumulate + 0 cube a inc b prime))


#### b.
the product of all the positive even integers less than n to n:

In [3]:
cat 1.3/sum_even_integers_by_filtered_accumulate_by_iteration.scm

(load "1.3/filtered_accumulate_by_iteration.scm")

(define (identity x) x)
(define (inc n) (+ n 1))

(define (sum-even-integers n)
  (filtered-accumulate + 0 identity 1 inc n even?))


## 1.3.2 Constructing Procedures Using $lambda$

In using $sum$ as in Section 1.3.1, it seems terribly awkward to have to define trivial procedures such as $pi$-$term$ and $pi$-$next$ just so we can use them as arguments to our higher-order procedure. Rather than define $pi$-$next$ and $pi$-$term$, it would be more convenient to have a way to directly specify “the procedure that returns its input incremented by 4” and “the procedure that returns the reciprocal of its input times its input plus 2.” We can do this by introducing the special form $lambda$, which creates procedures. Using $lambda$ we can describe what we want as

and

Then our $pi$-$sum$ procedure can be expressed without defining any auxiliary procedures as

In [4]:
cat 1.3/sum_pi_by_lambda_by_recursion.scm

(load "1.3/sum_by_recursion.scm")

(define (pi-sum a b)
  (sum (lambda (x) (/ 1.0 (* x (+ x 2))))
       a
       (lambda (x) (+ x 4))
       b))


Again using $lambda$, we can write the $integral$ procedure without having to define the auxiliary procedure $add$-$dx$:

In [6]:
cat 1.3/compute_integral_for_function_by_lambda.scm

(load "1.3/sum_by_recursion.scm")

(define (integral f a b dx)
  (* (sum f
          (+ a (/ dx 2.0))
          (lambda (x) (+ x dx))
          b)
     dx))


In general, $lambda$ is used to create procedures in the same way as $define$, except that no name is specified for the procedure:

The resulting procedure is just as much a procedure as one that is created using $define$.   
The only difference is that it has not been associated with any name in the environment. In fact,

is equivalent to

Like any expression that has a procedure as its value, a $lambda$ expression can be used as the operator in a combination such as

### Using $let$ to create $local$ $variables$

Another use of $lambda$ is in creating $local$ $variables$.   
We often need $local$ $variables$ in our procedures other than those that have been bound as $formal$ $parameters$.   
For example, suppose we wish to compute the function

$$f(x,y) = x(1 + xy)^2 + y(1 - y) + (1 + xy)(1 -y),$$

which we could also express as

$$a = 1 + xy,$$
$$b = 1 - y,$$
$$f(x,y) = xa^2 + yb + ab.$$

In writing a procedure to compute $f$ , we would like to include as $local$ $variables$ not only $x$ and $y$ but also the names of intermediate quantities like $a$ and $b$. One way to accomplish this is to use an auxiliary procedure to bind the $local$ $variables$:

Of course, we could use a $lambda$ expression to specify an anonymous procedure for binding our $local$ $variables$. The body of $f$ then becomes a single call to that procedure:

This construct is so useful that there is a special form called $let$ to make its use more convenient.  
Using $let$, the $f$ procedure could be written as

The general form of a $let$ expression is

The first part of the $let$ expression is a list of $name$-$expression$ pairs.  
When the $let$ is evaluated, each $name$ is associated with the value of the corresponding $expression$.   
The body of the $let$ is evaluated with these names bound as $local$ $variables$.   
The way this happens is that the $let$ expression is interpreted as an alternate syntax for

No new mechanism is required in the interpreter in order to provide $local$ $variables$.   
A $let$ expression is simply $syntactic$ $sugar$ for the underlying $lambda$ application

We can see from this equivalence that the scope of a variable specified by a $let$ expression is the body of the $let$.   
This implies that:

$\bullet$ $let$ allows one to bind variables as locally as possible to where they are to be used. For example, if the value of $x$ is 5, the value of the expression

is 38. Here, the $x$ in the body of the $let$ is 3, so the value of the let expression is 33.   
On the other hand, the $x$ that is the second argument to the outermost $+$ is still 5.

$\bullet$ The variables’values are computed outside the $let$. This matters when the expressions that provide the values for the $local$ $variables$ depend upon variables having the same names as the local variables themselves. For example, if the value of $x$ is 2, the expression

will have the value 12 because, inside the body of the $let$, $x$ will be 3 and $y$ will be 4 (which is the outer $x$ plus 2).

Sometimes we can use internal definitions to get the same effect as with $let$. For example, we could have defined the procedure $f$ above as

We prefer,however,to use $let$ in situations like this and to use $internal$ $define$ only for $internal$ $procedures$.

### Exercise 1.34: 
Suppose we define the procedure

Then we have

What happens if we (perversely) ask the interpreter to evaluate the combination $(f$ $f)$? Explain.